Euler's totient function #445

kushagraiitbbs opened the pull request:
### Have you read the [Contributing Guidelines on Pull Requests]
Yes

### Description
I have created Euler's_Totient_Function.cpp file in c++ folder and I have updated readme file.

### Checklist

- [x] I've read the contribution guidelines.
- [x] I've checked the issue list before deciding what to submit.
- [x] I've edited the `README.md` and link to my code.

## Related Issues or Pull Requests
#445 
(Write your answer here.)

kushagraiitbbs committed:
Creating Euler's_Totient_Code file

kushagraiitbbs committed:
Updating readme

atarax665 commented on C-Plus-Plus/math/Euler's_Totient_function.cpp:
```diff
@@ -0,0 +1,20 @@
+#include <bits/stdc++.h>
+using namespace std;
+int F(int n){
+    int arr[n+1],i,j;
```
It looks like you have copy pasted the code from somewhere. Try to code it yourself, follow proper indentation and refer to [C++ style guide](http://users.csc.calpoly.edu/~cstaley/General/CStyle.htm) to improve readability of your code. Make sure you go through all the 12 points mentioned in #398 .

kushagraiitbbs committed:
Update Euler's_Totient_function.cpp

Sir @atarax665 I have updated the function. Now F(n) calculates the number of coprime between 1 and n(where 1 and n are inclusive). Sir,I have indented it properly.

kushagraiitbbs commented:
@atarax665 Sir, I have updated the file. Now it is properly indented plus the function F(n) calculates the no. of coprimes between 1 and n (inclusive of 1 and n).

atarax665 commented on C-Plus-Plus/math/Euler's_Totient_function.cpp:
```diff
@@ -0,0 +1,26 @@
+    cin>>n;
+    cout<<F(n)<<endl; // always remember 0 is only coprime with 1
+    return 0;        
+}
```
Add sample Input/Output.

kushagraiitbbs replied:
@atarax665 Sir, I have added two sample cases at the end of code plus giving justification of dynamic programming

atarax665 commented on C-Plus-Plus/math/Euler's_Totient_function.cpp:
```diff
@@ -0,0 +1,26 @@
+#include <bits/stdc++.h>
```
Add description of Euler's Totient Function.

kushagraiitbbs replied:
@atarax665  Sir, I have added Euler's Totient Function description at the beginning of function code.

vgb-codes replied:
@atarax665 Just checking, would a ' in the programs name cause any issues?

atarax665 replied:
> @atarax665 Just checking, would a ' in the programs name cause any issues?

@VijayaGB98 No, it would be fine. Only spaces are encoded as %20 in URL.

siddharth25pandey reviewed (changes requested):
@kushagraiitbbs Please do the following modifications
* for user friendly interface, please display some message regarding the topic.

kushagraiitbbs committed:
Update Euler's_Totient_function.cpp

@atarax665 Sir, I have provided Euler's Totient Function in code plus I have added two sample inputs and giving justification of dynamic programmin

kushagraiitbbs committed:
Creating message at the top of code 

@siddharth25pandey @atarax665 Writing Euler's Totient Function definition at the beginning of code.

kushagraiitbbs commented:
> @kushagraiitbbs Please do the following modifications
> 
> * for user friendly interface, please display some message regarding the topic.

@siddharth25pandey I have added message of Euler's Totient Function at the beginning of code.

kushagraiitbbs commented:
@atarax665 @siddharth25pandey @VijayaGB98 I am a **HAKINCODE PARTICIPANT** (adding it for more clearity)

atarax665 commented on C-Plus-Plus/math/Euler's_Totient_function.cpp:
```diff
@@ -0,0 +1,48 @@
+// Euler's Totient Function represented by M(n)
+// For integer n it calculates the no. of coprimes
+// of n from 0 to n. Do refer line 18 for 
```
First of all Euler's totient function calculates the number of coprimes of **n** from **1 to n**. Which clearly means that all the numbers between 1 to n (both inclusive) that are relatively prime to **n** are to be only counted.
- The function written by you counts all the coprimes between 0 to n.

Please modify your function to produce required output.

kushagraiitbbs replied:
@atarax665 Sir, I have removed coprime pairs including 0 and I have updated the definition of Euler's Totient Function definition. So, I have changed sample output corresponding to it.

atarax665 replied:
> @atarax665 Sir, I have removed coprime pairs including 0 and I have updated the definition of Euler's Totient Function definition. So, I have changed sample output corresponding to it.

@kushagraiitbbs I would like to emphasis on **'relatively prime to n'**, which means only numbers that have common divisor as 1 with n are to be counted.

Example- 
Input = 4
Output = 2
Reason: Only gcd(1, 4) =1 and gcd(3, 4) = 1 will be counted.

Refer to [this](https://www.geeksforgeeks.org/eulers-totient-function/) for better insight.

kushagraiitbbs replied:
@atarax665 Sir, actually this is the very first code that I have written but you said that answer for 4,6 and 9 are wrong. So, I changed my code to find number of coprime pairs from 1 to n because this is the statement of my issue. Please do check my very first commit.

kushagraiitbbs replied:
@atarax665 This is the link of my very first commit 
https://github.com/TesseractCoding/NeoAlgo/pull/470/commits/b69525c78d250a2a8707cb20e7ea1c8e565ea8ed

atarax665 replied:
> @atarax665 Sir, actually this is the very first code that I have written but you said that answer for 4,6 and 9 are wrong. So, I changed my code to find number of coprime pairs from 1 to n because this is the statement of my issue. Please do check my very first commit.

My bad, actually I was reviewing this PR and #487  together and the function was wrong for #487 . I guess I commented the same on your PR too (Though I edited it later). Can you please restore the function ass in your first commit.

kushagraiitbbs replied:
@atarax665 Sir, I have restored my code with proper indentation and clear definition of Euler's Totient Function at the beginning of code plus sample input and output.

atarax665 replied:
Great! Do some minor changes and then it would be good to merge.

kushagraiitbbs committed:
Updating Euler's Totient function definition modifying code

@atarax665 Sir, I have updated function and Euler's Totient Function definition. Now, I have removed cases coprime pairs including 0.

kushagraiitbbs committed:
Modifying Euler's Totient Function

@atarax665 I have updated the function to start from 1 plus I have removed (0,1) pair and added (1,1)

kushagraiitbbs committed:
Restoring Euler's_Totient_Function.cpp

I am restoring my code with proper indentation and clear definition of Euler's Totient Function at the beginning of code plus sample input and sample output.

atarax665 commented on C-Plus-Plus/math/Euler's_Totient_function.cpp:
```diff
@@ -1,49 +1,37 @@
+    return arr[n];  //  here each arr[i] is giving number of coprimes of i from 1 to i
 }
 int main(){
     int n;
```
Add cout statement prompting user to enter a number. Do proper spacing to make the code more readable. (Add space between operators and operands, commas, and semicolons, and colons. Add space on the outside of parentheses and brackets)

atarax665 replied:
I have corrected the description of Euler's Totient Function in Issue #445 . Other contributors may get confused.

kushagraiitbbs replied:
@atarax665 Sir, I  have added spaces in opearators and operands. I have added comment on input statement too, so that they get a clear view of entering value.

kushagraiitbbs committed:
Update Euler's_Totient_function.cpp

Adding spaces between operators and operands to make it clear.

atarax665 commented on C-Plus-Plus/math/Euler's_Totient_function.cpp:
```diff
@@ -0,0 +1,38 @@
+using namespace std;
+int F(int n){
+    int arr[n + 1], i, j;
+    arr[0]=0, arr[1]=1;     // as 1 is coprime with itself
```
Add space between operator and operand in line 7, 8, 9, 11, 12, 13. Add cout<<"Enter a number : "; after line 22.

kushagraiitbbs replied:
@atarax665 Updated file, added cout<<"Enter a integer: ", added space between operator and operand as mentioned in above lines.

atarax665 replied:
Voila! LGTM :+1:

kushagraiitbbs committed:
Update Euler's_Totient_function.cpp

Adding spaces in operator and operand in line 7,8,9,11,12,13 and adding cout<<"Enter a integer: " in line 22

atarax665 commented on C-Plus-Plus/math/Euler's_Totient_function.cpp:
```diff
@@ -0,0 +1,39 @@
+
+int main(){
+    int n;
+    cout<<"Enter a integer: " ;
```
```suggestion
    cout<<"Enter an integer: " ;
```

siddharth25pandey reviewed (approved):
@kushagraiitbbs Looks fine to me.
Able to merge.
Thanks

iamrajiv commented:
@all-contributors please add @kushagraiitbbs for code
